code/game/machinery/bots/draculabot.dm

```diff
@@ -0,0 +1,228 @@
+
+/obj/machinery/bot/bloodbot/attack_hand(mob/user as mob)
+	if(..())
+		return
+	var/dat
+	dat += "<TT><B>Acula-class Blood Donation Bot v0.1</B></TT><BR><BR>"
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		var/datum/reagent/blood/B = H.get_blood(H.vessel)
+		if(B.data["virus2"])
+			dat += "<span class='danger'>WARNING: Viral agent detected. Ineligable for blood donation.</span><BR>"
+		else
+			dat += {"Welcome [H]! Your blood level is [round(B.volume/560*100)]%, and your blood type is [B.data["blood_type"]].<BR>
+				You must have at least [round(509/560*100)]% to donate blood. <a href='?src=\ref[src];donate=[1]'>Donate now!</a><BR>
+				The next reward will be dispensed after [55 - since_last_reward] units of blood are donated.<BR>"}
+	dat += {"Status: <A href='?src=\ref[src];power=1'>[src.on ? "On" : "Off"]</A><BR>"
+		Maintenance panel panel is [src.open ? "opened" : "closed"]<BR>"}
+	if(!src.locked || issilicon(user))
+		dat += "<TT>Blood Storage: "
+		var/counter = 0
+		if(!contents.len)
+			dat += "There are no blood packs available."
+		for (var/obj/item/weapon/reagent_containers/blood/E in contents)
+			counter++
+			dat += "Slot [counter]: [E] <A href='?src=\ref[src];slot=\ref[E]'>(Eject)</A><BR>"
+		dat += "The speaker switch is [src.quiet ? "off" : "on"]. <a href='?src=\ref[src];togglevoice=[1]'>Toggle</a><br>"
+
+	user << browse("<HEAD><TITLE>Dr. Acula Controls</TITLE></HEAD>[dat]", "window=autoblood")
+	onclose(user, "autoblood")
+	return
+
+/obj/machinery/bot/bloodbot/Topic(href, href_list)
+	if(..())
+		return
+	usr.set_machine(src)
+	add_fingerprint(usr)
+	if(href_list["power"] && allowed(usr))
+		if(on)
+			turn_off()
+		else
+			turn_on()
+
+	else if(href_list["slot"] && contents.len)
+		var/obj/item/weapon/reagent_containers/blood/E = locate(href_list["slot"])
+		E.forceMove(get_turf(src))
+		usr.put_in_hands(E) //Try to put it in the user's hands if available.
+		updateUsrDialog()
+
+	else if(href_list["donate"])
+		if(currently_drawing_blood)
+			speak("Sorry! One customer at a time!")
+			return
+		else
+			speak(pick("I'll just take a quick bite.","You may feel a slight sting.","There will be no anesthetic.","Delicious!","Don't mind if I do...","Thanks!"))
+			drink(usr)
+
+	else if(href_list["togglevoice"] && (!src.locked || issilicon(usr)))
+		quiet = !quiet
+
+	src.updateUsrDialog()
+
+/obj/machinery/bot/bloodbot/attackby(obj/item/weapon/W, mob/user)
+	if (istype(W, /obj/item/weapon/card/id)||istype(W, /obj/item/device/pda))
+		if (allowed(user) && !open && !emagged)
+			locked = !locked
+			to_chat(user, "<span class='notice'>Controls are now [locked ? "locked" : "unlocked"].</span>")
+			updateUsrDialog()
+		else
+			if(emagged)
+				to_chat(user, "<span class='warning'>ERROR</span>")
+			if(open)
+				to_chat(user, "<span class='warning'>Please close the access panel before locking it.</span>")
+			else
+				to_chat(user, "<span class='warning'>Access denied.</span>")
+
+	else
+		..()
```

Comment:
Iunno, honestly this whole proc is an artifact from when I stole the code from medbot. It's a little hard to see because I had to write so much from scratch, but this proc was pretty much left as the original. I'll look into it a little more closely when I do testing.